Storage/Rust/RustRemoteTabs.swift

```diff
@@ -0,0 +1,173 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+import Shared
+@_exported import MozillaAppServices
+
+private let log = Logger.syncLogger
+
+public extension RemoteTab {
+    func toRemoteTabRecord(client: RemoteClient) -> RemoteTabRecord {
+        let url = Foundation.URL(fileURLWithPath: self.urlHistory[0])
+        let history = self.urlHistory[1...].map { url in Foundation.URL(fileURLWithPath: url) }
+        let icon = self.icon != nil ? Foundation.URL(fileURLWithPath: self.icon ?? "") : nil
+        
+        return RemoteTabRecord(clientGUID: client.guid, URL: url, title: self.title, history: history, lastUsed: Timestamp(self.lastUsed), icon: icon)
+    }
+}
+
+public extension ClientRemoteTabs {
+    func toClientAndTabs(client: RemoteClient) -> ClientAndTabs {
+        return ClientAndTabs(client: client, tabs: self.remoteTabs.map { $0.toRemoteTabRecord(client: client) })
+    }
+}
+
+public class RustRemoteTabs {
+    let databasePath: String
+
+    let queue: DispatchQueue
+    var storage: TabsStorage?
+
+    private(set) var isOpen = false
+
+    private var didAttemptToMoveToBackup = false
+    
+    public init(databasePath: String) {
+        self.databasePath = databasePath
+        
+        queue = DispatchQueue(label: "RustRemoteTabs queue: \(databasePath)", attributes: [])
+    }
+    
+    private func open() -> NSError? {
+        do {
+            storage = try TabsStorage(databasePath: databasePath)
+            isOpen = true
+            return nil
+        } catch let err as NSError {
+            Sentry.shared.sendWithStacktrace(message: "Unknown error when opening Tabs Places database", tag: SentryTag.rustRemoteTabs, severity: .error, description: err.localizedDescription)
+
+            return err
+        }
+    }
+    
+    private func close() -> NSError? {
+        storage = nil
+        isOpen = false
+        return nil
+    }
+
+    public func reopenIfClosed() -> NSError? {
+        var error: NSError?
+
+        queue.sync {
+            guard !isOpen else { return }
+
+            error = open()
+        }
+
+        return error
+    }
+
+    public func forceClose() -> NSError? {
+        var error: NSError?
+
+        queue.sync {
+            guard isOpen else { return }
+
+            error = close()
+        }
+
+        return error
+    }
+    
+    public func sync(unlockInfo: SyncUnlockInfo) -> Success {
+        let deferred = Success()
+
+        queue.async {
+            guard self.isOpen else {
+                let error = TabsError.OpenDatabaseError(message: "Database is closed")
+                deferred.fill(Maybe(failure: error as MaybeErrorType))
+                return
+            }
+
+            do {
+                try _ = self.storage?.sync(unlockInfo: unlockInfo)
```

Comment:
This doesn't exist yet, right? Ditto for the reset case below - ie, we will need to add these functions to app-services, right?